### PR TITLE
Support for the ESP-IDF framework

### LINKED ISSUE #379: Support ESP-IDF Rust Target (for ESP32-C3 especially)
Right now it fails to build:
```
   Compiling socket2 v0.4.7
error[E0432]: unresolved import `libc::SOCK_RAW`
  --> C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\socket2-0.4.7\src\sys\unix.rs:64:16
   |
64 | pub(crate) use libc::SOCK_RAW;
   |                ^^^^^^^^^^^^^^ no `SOCK_RAW` in the root

error[E0432]: unresolved import `libc::SOCK_SEQPACKET`
  --> C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\socket2-0.4.7\src\sys\unix.rs:66:16
   |
66 | pub(crate) use libc::SOCK_SEQPACKET;
   |                ^^^^^^^^^^^^^^^^^^^^ no `SOCK_SEQPACKET` in the root

error[E0432]: unresolved import `libc::IP_HDRINCL`
  --> C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\socket2-0.4.7\src\sys\unix.rs:79:16
   |
79 | pub(crate) use libc::IP_HDRINCL;
   |                ^^^^^^^^^^^^^^^^ no `IP_HDRINCL` in the root

error[E0432]: unresolved import `libc::IP_RECVTOS`
  --> C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\socket2-0.4.7\src\sys\unix.rs:89:16
   |
89 | pub(crate) use libc::IP_RECVTOS;
   |                ^^^^^^^^^^^^^^^^ no `IP_RECVTOS` in the root

error[E0432]: unresolved imports `libc::ip_mreq_source`, `libc::IP_ADD_SOURCE_MEMBERSHIP`, `libc::IP_DROP_SOURCE_MEMBERSHIP`
   --> C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\socket2-0.4.7\src\sys\unix.rs:117:5
    |
117 |     ip_mreq_source as IpMreqSource, IP_ADD_SOURCE_MEMBERSHIP, IP_DROP_SOURCE_MEMBERSHIP,
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^  ^^^^^^^^^^^^^^^^^^^^^^^^  ^^^^^^^^^^^^^^^^^^^^^^^^^ no `IP_DROP_SOURCE_MEMBERSHIP` in the root
    |     |                               |
    |     |                               no `IP_ADD_SOURCE_MEMBERSHIP` in the root
    |     no `ip_mreq_source` in the root
    |
help: a similar name exists in the module
    |
117 |     ip_mreq_source as IpMreqSource, IP_ADD_MEMBERSHIP, IP_DROP_SOURCE_MEMBERSHIP,
    |                                     ~~~~~~~~~~~~~~~~~
help: a similar name exists in the module
    |
117 |     ip_mreq_source as IpMreqSource, IP_ADD_SOURCE_MEMBERSHIP, IP_DROP_MEMBERSHIP,
    |                                                               ~~~~~~~~~~~~~~~~~~

error[E0433]: failed to resolve: use of undeclared type `IovLen`
   --> C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\socket2-0.4.7\src\sys\unix.rs:790:38
    |
790 |     msg.msg_iovlen = min(bufs.len(), IovLen::MAX as usize) as IovLen;
    |                                      ^^^^^^ use of undeclared type `IovLen`

error[E0433]: failed to resolve: use of undeclared type `IovLen`
   --> C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\socket2-0.4.7\src\sys\unix.rs:850:38
    |
850 |     msg.msg_iovlen = min(bufs.len(), IovLen::MAX as usize) as IovLen;
    |                                      ^^^^^^ use of undeclared type `IovLen`

error[E0531]: cannot find unit struct, unit variant or constant `SOCK_RAW` in crate `libc`
   --> C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\socket2-0.4.7\src\sys\unix.rs:362:11
    |
362 |     libc::SOCK_RAW,
    |           ^^^^^^^^ not found in `libc`

error[E0531]: cannot find unit struct, unit variant or constant `SOCK_RDM` in crate `libc`
   --> C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\socket2-0.4.7\src\sys\unix.rs:364:11
    |
364 |     libc::SOCK_RDM,
    |           ^^^^^^^^ not found in `libc`

error[E0531]: cannot find unit struct, unit variant or constant `SOCK_SEQPACKET` in crate `libc`
   --> C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\socket2-0.4.7\src\sys\unix.rs:365:11
    |
365 |     libc::SOCK_SEQPACKET,
    |           ^^^^^^^^^^^^^^ not found in `libc`

error[E0412]: cannot find type `IovLen` in this scope
   --> C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\socket2-0.4.7\src\sys\unix.rs:790:63
    |
790 |     msg.msg_iovlen = min(bufs.len(), IovLen::MAX as usize) as IovLen;
    |                                                               ^^^^^^ not found in this scope

error[E0412]: cannot find type `IovLen` in this scope
   --> C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\socket2-0.4.7\src\sys\unix.rs:850:63
    |
850 |     msg.msg_iovlen = min(bufs.len(), IovLen::MAX as usize) as IovLen;
    |                                                               ^^^^^^ not found in this scope

error[E0412]: cannot find type `ip_mreqn` in crate `libc`
    --> C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\socket2-0.4.7\src\sys\unix.rs:1035:12
     |
1035 |   ) -> libc::ip_mreqn {
     |              ^^^^^^^^ help: a struct with a similar name exists: `ip_mreq`
     |
    ::: C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\libc-0.2.135\src\unix\newlib\mod.rs:51:1
     |
51   | / s! {
52   | |     // The order of the `ai_addr` field in this struct is crucial
53   | |     // for converting between the Rust and C types.
54   | |     pub struct addrinfo {
...    |
228  | |     }
229  | | }
     | |_- similarly named struct `ip_mreq` defined here

error[E0422]: cannot find struct, variant or union type `ip_mreqn` in crate `libc`
    --> C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\socket2-0.4.7\src\sys\unix.rs:1037:75
     |
1037 |           crate::socket::InterfaceIndexOrAddress::Index(interface) => libc::ip_mreqn {
     |                                                                             ^^^^^^^^ help: a struct with a similar name exists: `ip_mreq`
     |
    ::: C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\libc-0.2.135\src\unix\newlib\mod.rs:51:1
     |
51   | / s! {
52   | |     // The order of the `ai_addr` field in this struct is crucial
53   | |     // for converting between the Rust and C types.
54   | |     pub struct addrinfo {
...    |
228  | |     }
229  | | }
     | |_- similarly named struct `ip_mreq` defined here

error[E0422]: cannot find struct, variant or union type `ip_mreqn` in crate `libc`
    --> C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\socket2-0.4.7\src\sys\unix.rs:1042:77
     |
1042 |           crate::socket::InterfaceIndexOrAddress::Address(interface) => libc::ip_mreqn {
     |                                                                               ^^^^^^^^ help: a struct with a similar name exists: `ip_mreq`
     |
    ::: C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\libc-0.2.135\src\unix\newlib\mod.rs:51:1
     |
51   | / s! {
52   | |     // The order of the `ai_addr` field in this struct is crucial
53   | |     // for converting between the Rust and C types.
54   | |     pub struct addrinfo {
...    |
228  | |     }
229  | | }
     | |_- similarly named struct `ip_mreq` defined here

error[E0063]: missing field `sin_len` in initializer of `sockaddr_in`
   --> C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\socket2-0.4.7\src\sockaddr.rs:225:27
    |
225 |         let sockaddr_in = sockaddr_in {
    |                           ^^^^^^^^^^^ missing `sin_len`

error[E0063]: missing field `sin6_len` in initializer of `sockaddr_in6`
   --> C:\Users\steve\.cargo\registry\src\github.com-1ecc6299db9ec823\socket2-0.4.7\src\sockaddr.rs:260:28
    |
260 |         let sockaddr_in6 = sockaddr_in6 {
    |                            ^^^^^^^^^^^^ missing `sin6_len`

Some errors have detailed explanations: E0063, E0412, E0422, E0432, E0433, E0531.
For more information about an error, try `rustc --explain E0063`.
error: could not compile `socket2` due to 17 previous errors
```

### PR DESCRIPTION
**Pull request by jasta**
Smoke tested on esp32c3 dev board.  I've also tested a similar patch backported to v0.4.9 with much greater functionality including tokio + mio with other patches I've been working on and it's fully working.

Closes #379

### CODE REVIEW

**Comment by jasta** [comment]
I'm not sure how to make the cargo check tests work either, pointers greatly appreciated.

**Comment by jasta** [comment]
The clippy check that's failing appears to be talking about a file I didn't touch (tests/socket.rs)?  Weird...

File: src/sys/unix.rs [hunk 1]
```diff
@@ -539,7 +549,14 @@ impl RecvFlags {
     ///
     /// [`SEQPACKET`]: Type::SEQPACKET
     pub const fn is_end_of_record(self) -> bool {
-        self.0 & libc::MSG_EOR != 0
+        // TODO: Expose this constant in libc for the ESP-IDF/LwIP framework
+        #[cfg(target_os = "espidf")]
+        const MSG_EOR: libc::c_int = 8;
```
Comment (Thomasdezeeuw) [inline src/sys/unix.rs:554]: I don't want to maintain these constants so please add them to libc instead.

File: src/sys/unix.rs [hunk 2]
```diff
@@ -111,8 +111,13 @@ pub(crate) use libc::{
...
 };
 // Used in `RecvFlags`.
+#[cfg(not(any(target_os = "redox", target_os = "espidf")))]
+pub(crate) use libc::MSG_TRUNC;
 #[cfg(not(target_os = "redox"))]
-pub(crate) use libc::{MSG_TRUNC, SO_OOBINLINE};
+pub(crate) use libc::SO_OOBINLINE;
+// TODO: Expose in libc for ESP-IDF/LwIP
+#[cfg(target_os = "espidf")]
+pub(crate) const MSG_TRUNC: libc::c_int = 4;
```
Comment (Thomasdezeeuw) [inline src/sys/unix.rs:120]: I don't want to maintain these constants so please add them to libc instead.

**Review by Thomasdezeeuw — COMMENTED** [review]
You can ignore the Clippy warning

File: src/sys/unix.rs (hunk 1 above)
Comment (jasta) [reply src/sys/unix.rs:554]: I think we might have alternatives here, let me do a little digging and get back to you with options.

Comment (jasta) [reply src/sys/unix.rs:554]: Oops yeah we can follow redox's example here and just conditionally remove the feature.  I'll work on that soon and update.  Thanks for the review!

**Comment by jasta** [comment]
Much better, should be good to go now.

**Review by Thomasdezeeuw — APPROVED** [review]

**Comment by Thomasdezeeuw** [comment]
Thanks @jasta.

I've opened https://github.com/rust-lang/socket2/issues/454 to track CI support so we can ensure things keep working, any feedback on that would be welcome.

**Pull request merged**